### ISSUE TITLE
Unsigned to signed conversion error in SnmpPacket.marshalPDU()

### ISSUE THREAD

**Issue by reimda** [gosnmp/gosnmp#293]
SnmpPacket.MaxRepetitions is type uint8 throughout gosnmp.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L54

In SnmpPacket.marshalPDU(), MaxRepetitions is used to build a GetBulkRequest PDU where a signed value is expected.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L539

In this code, we're writing in ASN.1 BER (basic encoding rules).  The first byte is the data type, with 2 meaning signed integer, then the second byte is length of data in bytes, finally the data is one byte and comes from SnmpPacket.MaxRepetitions.

This unsigned to signed conversion error causes some agents (including net-snmp) to refuse to respond to the request.  Values 0 through 127 work fine in gosnmp because their signed representation is the same.  Unsigned values 128 through 255 don't fit in a signed 8 bit integer.  When they are misinterpreted as signed data they become -128 through -1.  Those negative values are invalid so an agent can't know how to handle them.

The SNMP RFC defines this PDU, the types of the data and their valid ranges.  Note that the INTEGER type here always means signed.
https://tools.ietf.org/html/rfc3416#section-3:
```
   max-bindings INTEGER ::= 2147483647
...
   BulkPDU ::=                         -- must be identical in
       SEQUENCE {                      -- structure to PDU
           request-id      INTEGER (-214783648..214783647),
           non-repeaters   INTEGER (0..max-bindings),
           max-repetitions INTEGER (0..max-bindings),

           variable-bindings           -- values are ignored
               VarBindList
       }
```

ASN.1 BER is defined in ITU-T X.690.  It's available at https://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf.  Section 8.3 describes how to encode an integer.

There are other places where gosnmp uses an unsigned integer in a BER context that expects a signed integer.  I'm not sure if these cause similar failures, but they should be checked.
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L536
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L563
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/marshal.go#L560

**Comment by reimda** [comment]
This seems to be the cause of a telegraf issue.  https://github.com/influxdata/telegraf/issues/8823

**Comment by TimRots** [comment]
Thanks for the detailed report

> Values 0 through 127 work fine in gosnmp because their signed representation is the same. Unsigned values 128 through 255 don't fit in a signed 8 bit integer. When they are misinterpreted as signed data they become -128 through -1. Those negative values are invalid so an agent can't know how to handle them.

Would it be sufficient of a fix to wraparound if ``MaxRepetitions > 127`` @reimda @SuperQ ?
If so, a possible fix could look like:
```
buf.Write([]byte{2, 1, (packet.MaxRepetitions & 0x7F)})
```

test:
```
func main() {
	var MaxRepetitions uint8
	for MaxRepetitions = 0; MaxRepetitions < 255; MaxRepetitions++ {
		buf := new(bytes.Buffer)
		buf.Write([]byte{2, 1, (MaxRepetitions & 0x7F)})
		fmt.Printf("%d %b\n", (MaxRepetitions & 0x7F), buf.Bytes())
	}
}
123 [10 1 1111011]
124 [10 1 1111100]
125 [10 1 1111101]
126 [10 1 1111110]
127 [10 1 1111111]
0 [10 1 0]
1 [10 1 1]
2 [10 1 10]
3 [10 1 11]
4 [10 1 100]
```

**Comment by reimda** [comment]
Values greater than 127 can still be encoded in BER, they just take two data bytes.  Also, keep in mind that in the SNMP RFC max-bindings is valid from 0..2147483647.  If we're not willing to change gosnmp's API to allow the user to express that whole range, we should at least come up with a way to make 0..255 work and not just chop off the high bit.

There are functions in helper.go to marshal other types, see for example marshalInt32().  I'm not sure why we are writing the uint8 values directly with those buf.Write(...) lines.  Writing directly is bad idea.  The type and length bytes are magic numbers and require people reading the code in the future to be very familiar with ASN.1 BER.  It's just not very readable (and it probably contributed to this bug being around for so long.)

Even if we're going to keep MaxRepetitions as uint8 in gosnnp, we should still use a marshal function.  We might want to add marshalUint8() or modify an existing marshal function like marshalUint32().  That one already has the logic to make values larger than 127 into two bytes:
https://github.com/gosnmp/gosnmp/blob/9f003f366c564e2bd597dd765f5b97cdcee821e3/helper.go#L382

**Comment by TimRots** [comment]
With your comments in mind tried to implement the suggestions for max-repetitions.
Hoping you and @SuperQ have time review and discussion on how to best approach this fix in #295
